lib/Optimizer/Transforms/FactorQuantumAlloc.cpp

```diff
@@ -42,21 +42,53 @@ class AllocaPat : public OpRewritePattern<quake::AllocaOp> {
     for (std::size_t i = 0; i < size; ++i)
       newAllocs.emplace_back(rewriter.create<quake::AllocaOp>(loc, refTy));
 
-    // 2. Visit all users and replace them accordingly.
-    for (auto *user : allocOp->getUsers()) {
-      if (auto dealloc = dyn_cast<quake::DeallocOp>(user)) {
-        rewriter.setInsertionPoint(dealloc);
-        auto deloc = dealloc.getLoc();
-        for (std::size_t i = 0; i < size - 1; ++i)
-          rewriter.create<quake::DeallocOp>(deloc, newAllocs[i]);
-        rewriter.replaceOpWithNewOp<quake::DeallocOp>(dealloc,
-                                                      newAllocs[size - 1]);
-        continue;
+    std::function<LogicalResult(Operation *, std::int64_t)> rewriteOpAndUsers =
+        [&](Operation *op, std::int64_t start) -> LogicalResult {
+      // First handle the users. Note that this can recurse.
+      for (auto *user : op->getUsers()) {
+        if (auto dealloc = dyn_cast<quake::DeallocOp>(user)) {
+          rewriter.setInsertionPoint(dealloc);
+          auto deloc = dealloc.getLoc();
+          for (std::size_t i = 0; i < size - 1; ++i)
+            rewriter.create<quake::DeallocOp>(deloc, newAllocs[i]);
+          rewriter.replaceOpWithNewOp<quake::DeallocOp>(dealloc,
+                                                        newAllocs[size - 1]);
+          continue;
+        }
+        if (auto subveq = dyn_cast<quake::SubVeqOp>(user)) {
+          std::int64_t lowInt = -1;
+          if (auto low =
+                  dyn_cast<arith::ConstantOp>(subveq.getLow().getDefiningOp()))
+            if (auto myInt = dyn_cast<IntegerAttr>(low.getValue()))
+              lowInt = start + myInt.getInt();
+          if (lowInt < 0) {
+            subveq.emitOpError("Invalid subveq operator found");
```

Comment:
This pass should not fail with hard errors here. A subveq operation need not have constant bounds and this pass should not be forced to only run when there are constant bounds.

Reply:
The changes in the `runAnalysis` function have checks to ensure that it has constant bounds. If it does not have constant bounds, the original alloca op will not be added to the list of ops that needs to be rewritten. This is a check to ensure that the op hasn't snuck into the `matchAndRewrite` function after the analysis should've silently ignored that operation. This check should not fail under normal operation, unless somebody updates the `runAnalysis` function in the future without realizing they might break downstream code; this is a guard against that scenario.

Reply:
That being said - I can still remove the `emitOpError` call.

Reply:
Updated in e9a614cec